app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -582,8 +668,39 @@ public void reportDownloadProgress(int totalRead) {
         publishProgress(DataPullTask.PROGRESS_DOWNLOADING, totalRead);
     }
 
+    /**
+     * In order to achieve the impression of the progress bar updating smoothly during the wait
+     * period after a retry response is received, use a timer task to report incremental progress
+     * from the state we were last in to the progress value that the server just reported back to
+     * us, splitting up the wait time into uniform time periods for reporting each unit of progress.
+     */
+    private void startReportingServerProgress() {
+        long millisUntilNextAttempt = retryAtTime - System.currentTimeMillis();
+        int amountOfProgressToCoverThisCycle = serverProgressCompletedSoFar - lastReportedServerProgressValue;
+        if (amountOfProgressToCoverThisCycle == 0) {
+            return;
+        }
+        long intervalAllottedPerProgressUnit = millisUntilNextAttempt / amountOfProgressToCoverThisCycle;
+
+        final Timer reportServerProgressTimer = new Timer();
+        reportServerProgressTimer.schedule(new TimerTask() {
+                @Override
+                public void run() {
+                    if (lastReportedServerProgressValue == serverProgressCompletedSoFar) {

```

Comment:
Not sure how important it is to synchronize around modifying and reading `serverProgressCompletedSoFar` and `serverProgressTotal`